internal/controllers/machine/drain/drain.go

```diff
@@ -0,0 +1,353 @@
+				break
+			}
+			if status.Reason == PodDeleteStatusTypeWarning {
+				deleteWarnings = append(deleteWarnings, status.Message)
+			}
+		}
+
+		// Note: It only makes sense to aggregate warnings if we are going ahead with the deletion.
+		// If we don't, it's absolutely fine to just use the status from the filter that decided that
+		// we are not going to delete the Pod.
+		if status.Delete &&
+			(status.Reason == PodDeleteStatusTypeOkay || status.Reason == PodDeleteStatusTypeWarning) &&
+			len(deleteWarnings) > 0 {
+			status.Reason = PodDeleteStatusTypeWarning
+			status.Message = strings.Join(deleteWarnings, ", ")
+		}
+
+		// Add the pod to PodDeleteList no matter what PodDeleteStatus is,
+		// those pods whose PodDeleteStatus is false like DaemonSet will
+		// be caught by list.errors()
+		pod.Kind = "Pod"
+		pod.APIVersion = "v1"
+		pods = append(pods, PodDelete{
+			Pod:    pod,
+			Status: status,
+		})
+	}
+	list := &PodDeleteList{items: pods}
+	return list
+}
+
+// EvictPods evicts the pods.
+func (d *Helper) EvictPods(ctx context.Context, podDeleteList *PodDeleteList) EvictionResult {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Sort podDeleteList, this is important so we always deterministically evict Pods and build the EvictionResult.
+	// Otherwise the condition could change with every single reconcile even if nothing changes on the Node.
+	sort.Slice(podDeleteList.items, func(i, j int) bool {
+		return fmt.Sprintf("%s/%s", podDeleteList.items[i].Pod.GetNamespace(), podDeleteList.items[i].Pod.GetName()) <
+			fmt.Sprintf("%s/%s", podDeleteList.items[j].Pod.GetNamespace(), podDeleteList.items[j].Pod.GetName())
+	})
+
+	var podsToTriggerEviction []PodDelete
+	var podsWithDeletionTimestamp []PodDelete
+	var podsToBeIgnored []PodDelete
+	for _, pod := range podDeleteList.items {
+		switch {
+		case pod.Status.Delete && pod.Pod.DeletionTimestamp.IsZero():
+			podsToTriggerEviction = append(podsToTriggerEviction, pod)
+		case pod.Status.Delete:
+			podsWithDeletionTimestamp = append(podsWithDeletionTimestamp, pod)
+		default:
+			podsToBeIgnored = append(podsToBeIgnored, pod)
+		}
+	}
+
+	log.V(2).Info("Drain not completed yet, there are still Pods on the Node that have to be drained",
+		"podsToTriggerEviction", stringPodDeleteList(podsToTriggerEviction),
+		"podsWithDeletionTimestamp", stringPodDeleteList(podsWithDeletionTimestamp),
+	)
+
+	// Trigger evictions for at most 10s. We'll continue on the next reconcile if we hit the timeout.
+	globalTimeout := 10 * time.Second
+	ctx, cancel := context.WithTimeout(ctx, globalTimeout)
+	defer cancel()
+
+	res := EvictionResult{
+		PodsFailedEviction: map[string][]*corev1.Pod{},
+	}
+
+	for _, pd := range podsToBeIgnored {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pd.Pod))
+		if pd.Status.Reason == PodDeleteStatusTypeWarning && pd.Status.Message != "" {
+			log = log.WithValues("reason", pd.Status.Message)
+		}
+
+		log.V(3).Info("Skip evicting Pod because it should be ignored")
```

Comment:
nit
```suggestion
		log.V(4).Info("Skip evicting Pod because it should be ignored")
```
Same for all the trace messages for a specific pod

Reply:
I was reading our guidelines and trying to align to previous usage.

We have many logs in the cluster topolgoy controller that looks like this:
```
log.V(3).Infof("No changes for %s", tlog.KObj{Obj: in.desired})
```

In which way is this log here different? (I'm trying to find a way to figure out the log levels correctly but I'm seriously struggling)

Reply:
```go
- Logs at the lower levels of verbosity (<=3) are meant to document “what happened” by describing how an object status
  is being changed by controller/reconcilers across subsequent reconciliations; as a rule of thumb, it is reasonable
  to assume that a person reading those logs has a deep knowledge of how the system works, but it should not be required
  for those persons to have knowledge of the codebase. 
- Logs at higher levels of verbosity (>=4) are meant to document “how it happened”, providing insight on thorny parts of
  the code; a person reading those logs usually has deep knowledge of the codebase. 
```

For me this is still part of "what happend' (or in this case what didn't happen), and not part of "how it happened"

Reply:
I'm not blocking on this, just providing a PoV based on my take on guidelines:

We are in the machine reconciler, so when I'm looking at "what happened", is what happened to the machine.

To explain "what happened to the machine" (verbosity <=3), I don't think it it is not necessary to have detail to what happen to each pod on the machines, but it should be enough to just a summary of this that surfaces why machine drain is blocked, with possibly a list of few Pods blocking (and we already have this outside of this func). 
 
Instead, when I'm looking at “how it happened” (verbosity >=4), it s ok to get details to what happen to each pod on the machines, because this allows to dive in into on why drain is blocked

Looking at this from another angle, I fear that logging details about what happen to each pod, would be noisy, so I would prefer to have it only at higher level of verbosity

But again, not blocking on this, just providing another PoV

Reply:
Sounds okay, let's go with 4. I don't have a strong opinion if it should be 3 or 4. Your reasoning sounds fine.

Mostly just trying to get to a point where I can pick the correct log levels easiliy because I have a good understanding of the log levels :)

Reply:
Changed to 4 will push in a bit